src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -478,31 +476,24 @@ public String getShortDescription() {
         /**
          * Something we can use to check abort and read permissions.
          * Normally this will be a {@link Run}.
-         * However if things are badly broken, for example if the build has been deleted,
+         * If that has been deleted, we can fall back to the {@link Job}.
+         * If things are badly broken, for example if the whole job has been deleted,
          * then as a fallback we use the Jenkins root.
          * This allows an administrator to clean up dead queue items and executor cells.
-         * TODO make {@link FlowExecutionOwner} implement {@link AccessControlled}
-         * so that an implementation could fall back to checking {@link Job} permission.
          */
         @NonNull
         @Override public ACL getACL() {
-            try {
-                if (!context.isReady()) {
-                    return Jenkins.get().getACL();
-                }
-                FlowExecution exec = context.get(FlowExecution.class);
-                if (exec == null) {
-                    return Jenkins.get().getACL();
-                }
-                Queue.Executable executable = exec.getOwner().getExecutable();
-                if (executable instanceof AccessControlled) {
-                    return ((AccessControlled) executable).getACL();
+            Run<?, ?> r = runForDisplay();
+            if (r != null) {
+                return r.getACL();
+            } else {
+                Jenkins j = Jenkins.get();
+                Job<?, ?> job = j.getItemByFullName(runId.substring(0, runId.lastIndexOf('#')), Job.class);
```

Comment:
Not sure if it is accurate or relevant here, but `getItemByFullName`'s Javadoc says it can throw `AccessDeniedException`. Maybe we should leave the outer `try/catch` in place just in case?

Reply:
Theoretically if you could `DISCOVER` but not `READ` the job, yes. In that case the GUI at least would presumably not be even showing the item (so no red **X** either) but yes we should be defensive here.